### ISSUE TITLE
Pybind vs Implementation in Python

### ISSUE THREAD

**Issue by spacey-sooty** [GrappleRobotics/libgrapplefrc#8]
@JaciBrunning I'd be curious to know your opinion between using Pybind over the native implementation or creating a in Python implementation of the LaserCAN class. I know something like [polars](https://pola.rs/) uses Rust bindings within Python similar to how the FFI is being used for C++ here, I'd be curious to know what you think the best approach is.

**Comment by virtuald** [comment]
I hear [pyo3](https://pyo3.rs/v0.20.0/) is great for making python libraries.

If you want to publish the resulting wheel so that the robotpy-installer can download it, the wheel needs to be published to artifactory.. but we can hook you up with an account.

**Comment by spacey-sooty** [comment]
> I hear [pyo3](https://pyo3.rs/v0.20.0/) is great for making python libraries.
> 
> If you want to publish the resulting wheel so that the robotpy-installer can download it, the wheel needs to be published to artifactory.. but we can hook you up with an account.

I had a quick look at that and it looks really good. I'll have a go using that to create it, then when it's just about ready I'll ping you about publishing it so the robotpy-installer can download it.

**Comment by JaciBrunning** [comment]
Great work @spacey-sooty 

I might see if there's some way we can leverage the existing C bindings we have just so it's more DRY, but I imagine that's neither here nor there. PyO3 does provide a very nice interface but I'm a little worried about the amount of extra build system effort it takes. 
@virtuald I'll have a peek at how we can go about packaging these for dist this weekend :)

**Comment by virtuald** [comment]
We use [robotpy-build](https://github.com/robotpy/robotpy-build) to wrap all the wpilib C++ libraries and all the other vendors with pybind11. It's relatively easy to use and fairly robust, so wrapping your C++ library is an easy alternative if you don't want to do the pure rust thing with PyO3.

**Comment by spacey-sooty** [comment]
Another option thats been getting some usage is [nanobind](https://github.com/wjakob/nanobind/). PhotonVision was considering using it, I know Sleipnir does. Its significantly faster than Pybind11

**Comment by JaciBrunning** [comment]
This should be closed now that we got PyO3 working :)

**Comment by virtuald** [comment]
The robotpy installer pulls roborio wheels from the wpilib artifactory repo (see https://github.com/wpilibsuite/docker-images/blob/main/cross-ubuntu-py/pip-aarch64.conf ... Peter can set you up with a token). The wheels have to end in `linux_roborio.whl`.

Just publish your x86 wheels to pypi like normal python project and the robotpy installer will pull those for simulation.

Additionally, you'll want to publish the package name in the wpilib docs, there's a table with python package names.

**Comment by Chronosera** [comment]
![Image](https://github.com/user-attachments/assets/4de8c362-2514-4bfd-902f-2b31a0b3a78e)

PyPi still shows the same version of libgrapplefrc-py (2025.0.2) as latest, and we are unable to successfully install the package successfully - screenshot of our pyproject.toml and console output when running a robotpy sync attached.

**Comment by virtuald** [comment]
They only uploaded wheels for Python 3.10 (Linux) or 3.12 (Windows), so it will only work for those interpreters (you appear to have 3.13).

@JaciBrunning you should build more wheels, including on macOS. Something that will help is if you build an abi3 wheel, it can be used on multiple python versions. PyO3's documentation indicates it's possible to do this, but I don't have any experience with PyO3.

**Comment by JaciBrunning** [comment]
OK I'll have a look shortly and keep you updated.

**Comment by JaciBrunning** [comment]
@Chronosera give it a shot now, the binaries should be uploaded properly

**Comment by Chronosera** [comment]
Looks to be working now!

**Comment by Chronosera** [comment]
The library is installing properly, but it appears to be empty when trying to import anything (ex. The LaserCAN or can_bridge_tcp).

**Comment by virtuald** [comment]
On Linux, it seems to have things in it:

```
>>> import libgrapplefrc
>>> dir(libgrapplefrc)
['LaserCAN', 'MitoCANdria', '__all__', '__builtins__', '__cached__', '__doc__', '__file__', '__loader__', '__name__', '__package__', '__path__', '__spec__', 'can_bridge_tcp', 'libgrapplefrc']
```

There's no `.pyi` file, so it's likely your IDE won't autocomplete any of that.

**Comment by Chronosera** [comment]
All we can see is __all__ when in the IDE looking at the package files - even without auto fill, we're getting errors that the class doesn't exist (in this case, can_bridge_tcp). Will get a snip of it later today!

**Comment by virtuald** [comment]
Yes, your IDE is going to show errors, but that doesn't necessarily mean there will be errors at runtime. Python and compiled extensions and IDEs play well together only if there are .pyi files, and there aren't any.

If there are errors at runtime, that's a different problem. Which OS?

**Comment by Chronosera** [comment]
So we are not having errors at runtime, and the code appears to work fine initially - however, we are unable to figure out how to actually get the LaserCAN readout - following based off of the Java code example have been unsuccessful, and with no autofill from the IDE we are completely lost.

**Comment by JaciBrunning** [comment]
You'll want the `get_measurement()` function. I'll see if I can get a .pyi file generated

**Comment by Chronosera** [comment]
So we have been doing further testing - when we use the get_measurement() function, it returns a "LaserCANMeasurement" object, not the distance - looking further into the code, we saw that using get_measurement().distance_mm should return the int we are looking for, but it reports back in runtime that there is no distance_mm attribute within "LaserCANMeasurement" when using get_measurement().

**Comment by JaciBrunning** [comment]
Ah, seems pyo3 didn't generate the getters for that class. Try release `2025.1.2` - it should have those properties on the measurement class now

**Comment by Chronosera** [comment]
Works great! Thank you for your help!
